### ISSUE TITLE
AffiliateWP core meta box class (AffWP_Metabox_Base)

### ISSUE THREAD

**Issue by ramiabraham** [AffiliateWP/AffiliateWP#1225]
PR: #1420
Related: #1076

Initially defined among commits around the 1.8 reports overhaul (#1076, branch: https://github.com/AffiliateWP/AffiliateWP/tree/issue/1076), this thread will be the home for the AffiliateWP core meta box class and related methods, in case we need to ship separately.

----

(Updated 2016-May-12):

All meta boxes in use in core are in the process of being replaced with instances of an `AffWP_Metabox_Base` class extension.

The `AffWP_Metabox_Base` class defines a base class for creating meta boxes anywhere in the AffiliateWP admin section.

An array of permissible screens in which meta boxes may be displayed (via `get_screen_id`) are defined in `AffWP_Metabox_Base->affwp_screen`.

The context is also available, and determines the column per `add_meta_box`. This option is defined in `AffWP_Metabox_Base->context`.

[An example of extending this base class can be seen here](https://github.com/AffiliateWP/AffiliateWP/blob/issue/1076/includes/admin/reports/class-metabox-affiliate-leaderboard.php).

**Comment by pippinsplugins** [comment]
This is looking great and definitely a good start.

To help with testing, I'd love to see all existing meta boxes that we register get converted over to this class. That will help test for any issues.

I'd also love to see this work for adding meta boxes to just about all (if not all) of our custom screens: reports, tools, etc.

@ramiabraham looks like you've added `do_action( 'affwp_meta_boxes' )` to the reports page. Right now, this would result in all meta boxes getting displayed there, so we'll need to put in page / context checking to ensure only metaboxes registered for that page / context are displayed.

**Comment by ramiabraham** [comment]
Right on. Regarding the Overview page, should the meta boxes within be converted (and remain in their current location)? I know removal of the Overview page is on the table.

Can continue moving them over to the `affiliate-wp-reports` page(s) (some are in already).

So far, I have:

- For the `affwp_meta_boxes ` action, I'll add a method that returns an array of screen IDs to ensure the correct context.

- Going to separate meta box related activity into a new branch, and ping this issue on related commits

- Add any adjustments needed, testing and examples for adding a meta box to each of our admin screens.

- String translation fix in base class init method

**Comment by pippinsplugins** [comment]
If we're going to do it, I'd like to do it 100%. All existing metaboxes should be converted.

**Comment by ramiabraham** [comment]
Further info on this:

The `AffWP_Metabox_Base` class defines a base class for creating meta boxes anywhere in the AffiliateWP admin section.

An array of permissible screens in which meta boxes may be displayed (via `get_screen_id`) are defined in `AffWP_Metabox_Base->affwp_screen`.

The context is also available, and determines the column per `add_meta_box`. This option is defined in `AffWP_Metabox_Base->context`.

[An example of extending this base class can be seen here](https://github.com/AffiliateWP/AffiliateWP/blob/issue/1076/includes/admin/reports/class-metabox-affiliate-leaderboard.php).

**Comment by pippinsplugins** [comment]
@ramiabraham Do we have this split off from #1076 or is everything in that issue branch? We had discussed this being in 1.9, but it looks like reporting overhaul is using it, which is slated for 1.8.

**Comment by ramiabraham** [comment]
I have a reset of all reports areas in 1.8 using this for #1076 locally, will push to #1076.

**Comment by ramiabraham** [comment]
_Static markup restored in https://github.com/AffiliateWP/AffiliateWP/commit/9be5a81ded5eafac25bbba184629a6b12f2669ed_

**Comment by pippinsplugins** [comment]
:+1:

**Comment by ramiabraham** [comment]
Implementation of AffiliateWP core meta boxes will be on branch [issue/1225 ](https://github.com/AffiliateWP/AffiliateWP/tree/issue/1225)

**Comment by ramiabraham** [comment]
Meta boxes are implemented for the Overview page.

The primary structural notes on meta boxes are:

- each meta box is a class, which extends the `AffWP_Metabox_Base` class.
- `AffWP_Metabox_Base` lives in `/includes`, an agnostic home which makes no assumption on the context of any actual meta box.
- The toggle and positioning of meta boxes are enabled via `postboxes.add_postbox_toggles in WP core.
- The allowed screen IDs are defined in `AffWP_Metabox_Base::affwp_screen`. As of now, it's an array of all screen IDs used by AffiliateWP.
- The location of meta boxes, relative to their context, is within a subdir of the context. So for reports, they'd live in `reports/metaboxes`, etc:

![meta box directory structure](http://i.imgur.com/aHTGd6S.png "meta box dir")


_TODO:_
- Aside from this, I can think only of the `reports` page tabs as needing meta boxes in AffWP core...are there any other areas anyone sees where a meta box should be used?
- #1076 (reports overhaul) does away with the `Overview` page altogether. With what's in place/planned for `1.9`, the meta boxes implemented above are going to be carried over to the primary landing tab for `Reports` entirely, and the `Overview` page will retire in favor of the `Reports` section.
- Just semantics, but I'll be changing any instance of "metabox" to "meta box", with the exception of dir names, as noted above with `reports/metaboxes` :)

**Comment by pippinsplugins** [comment]
This is looking great!

A few things I see:

1. Box reordering does not save. Is that handled by core automagically or do we need to add a save routine?

2. The action `affwp_reports_meta_boxes` is fired three times in `affwp_reports_tab_overview()`. That concerns me because someone might see that as a hook they can tie into, but doing so result in callbacks functions firing three times on the same page.

3. Is it possible to define our own contexts for `do_meta_boxes()`? `normal`, `side`, and `advanced` don't make as much sense for as as they do WP core.

4. Can we remove the need to have each meta box call `add_action( 'add_meta_box',     array( $this, 'add_meta_box' ) );` in `init()`? That should be able to be done automatically from the base class. The only thing that the subclasses should typically be defining is the position and the content.

**Comment by ramiabraham** [comment]
1. Reordering _should_ be handled by core automagically - probably a bug I introduced in screen options methods.

2. Good catch

3. No idea presently, but _I will find a way_ if not built into core already, because this rocks.

4. 
![yes](http://i.imgur.com/2p8qZzh.jpg "ywc")

**Comment by pippinsplugins** [comment]
Oh one more thing I noticed. The drag / zoom controls are still on the graphs:
![screen shot 2016-06-06 at 11 31 36 am](https://cloud.githubusercontent.com/assets/1034109/15829667/47018382-2bda-11e6-860b-99c996236e09.png)
Looks like they were left behind when this was pulled out of #1076

**Comment by ramiabraham** [comment]
What should the name of our various meta box contexts be?

**Comment by pippinsplugins** [comment]
`primary`, `secondary`, `tertiary`.

**Comment by ramiabraham** [comment]
Ready for review (except for screen options saving meta box positions; in progress).

**Comment by ramiabraham** [comment]
All set here; ready for review.

**Comment by ramiabraham** [comment]
A note on my removing the reports page meta boxes; since this is slated for 1.9, and reports stuff for 1.9 greatly changes the reports area, creating meta boxes for this area here in #issue/1225 would just be removed later on. I'll leave those alone for now.

This makes it so that the only places in which core meta boxes are being used (as of https://github.com/AffiliateWP/AffiliateWP/commit/21fba3e7fac7b79d506f62fb22a99346f911e6ac) is the Overview page.

**Comment by DrewAPicture** [comment]
@ramiabraham Can you create a proper PR? The only reference I can find is to a direct feature branch.

**Comment by ramiabraham** [comment]
All set: https://github.com/AffiliateWP/AffiliateWP/pull/1420

**Comment by ramiabraham** [comment]
_(Related proposal to automatically add meta box nonces to all admin pages: https://github.com/AffiliateWP/AffiliateWP/issues/1422)_

**Comment by pippinsplugins** [comment]
Looking good! Drag and drop with save now works perfectly.

A few things I've found:

1. There is still CSS left over from the reports issue branch: https://github.com/AffiliateWP/AffiliateWP/pull/1420/files#diff-d1caebe36bbae5520fb7ef26f2e7db1dR176

2. The `assets/js/jquery.flot.navigate.js` file is left over from the reports issue branch.

3. Metaboxes are output with `do_meta_boxes( 'toplevel_page_affiliate-wp', 'primary', null )`. Can we change `toplevel_page_affiliate-wp` or is the `toplevel_page_` part required by WP core? If we can, I'd prefer that be `affwp_meta_boxes_{page}`.

4. Changing the number of columns in screen options didn't work for me. It triggered a JS error: `Uncaught TypeError: jQuery.plot is not a function`.

5. Mentioned this [above](https://github.com/AffiliateWP/AffiliateWP/issues/1225#issuecomment-224006927); I'd like to improve on the `init()` method. Right now it looks like this:
```php
/**
 * Initialize
 *
 * @access  public
 * @return  void
 * @since   1.9
 */
public function init() {
	add_action( 'add_meta_box',              array( $this, 'add_meta_box' ) );
	add_action( 'affwp_overview_meta_boxes', array( $this, 'add_meta_box' ) );

	$this->meta_box_name = __( 'Most Valuable Affiliates', 'affiliate-wp' );
}
```

Both `add_action()` calls will be required by every metabox and can be handled by the base class. The `init()` method should only handle setting the name of the meta box and anything custom someone wants to do. We need to be proactive in preventing mistakes by not requiring developers remember to add the actions here.

It should look like this once complete:
```php
/**
 * Initialize
 *
 * @access  public
 * @return  void
 * @since   1.9
 */
public function init() {
	$this->meta_box_name = __( 'Most Valuable Affiliates', 'affiliate-wp' );
}
```

6. The meta box base class file should be loaded in the main plugin file so that it is available to developers without the need to manually load it.

**Comment by ramiabraham** [comment]
1, 2. Corrected

3. Yeah `toplevel_page_` is required by core; it appends stuff depending on context. for non-cpt plugin settings that have a top-level item defined, it does this (see other screen IDs, such as `toplevel_page_gf_edit_forms`)

4. I opted to _remove_ this option instead, if that works? Opted for this since the meta boxes auto-sort depending on screen size already.

5. Adjusted as noted, with minor tweaks. I abstracted the method entirely in the base (moving the two hooks to the base constructor), and defined an`$action` var, which changes the init method of all children to:
```
/**
 * Initialize.
 *
 * Define the meta box name,
 * and the action on which to hook the meta box here.
 *
 * Example:
 *
 * $this->action        = 'affwp_overview_meta_boxes';
 * $this->meta_box_name = __( 'Name of the meta box', 'affiliate-wp' );
 *
 * @access  public
 * @return  void
 * @since   1.9
 */
public function init() {
	$this->action                 = 'affwp_overview_meta_boxes';
	$this->meta_box_name = __( 'Most Valuable Affiliates', 'affiliate-wp' );
}```

**Comment by pippinsplugins** [comment]
3. Ok that's fine.

4. :+1:

5. Great, will review further later today.

**Comment by ramiabraham** [comment]
(As of https://github.com/AffiliateWP/AffiliateWP/commit/2b64760b746e0fc30d36972fdacc3926b65dea0a this branch is correctly reset against `release/1.9`)

**Comment by ramiabraham** [comment]
- namespaces added
- minor adjustment to inline docs formatting

**Comment by pippinsplugins** [comment]
Namespaces look good at first glance. Will test more tomorrow.

**Comment by pippinsplugins** [comment]
@ramiabraham Looks like something died in the JS. Dragging and opening / closing no longer works on the Overview page.

**Comment by pippinsplugins** [comment]
Looks good now.

The metabox base class is still loaded in overview.php. That should be removed so it's only loaded by the main plugin file.

I believe we can get rid of `require_once AFFILIATEWP_PLUGIN_DIR . 'includes/admin/overview/screen-options.php';` as well since we're not using screenshots anymore.

**Comment by ramiabraham** [comment]
- Removing that causes a class not found fatal for some reason, despite it being present in the main plugin file. Looking for the issue.

- Removing `includes/admin/overview/screen-options.php` will cause meta boxes not to open/close or be movable, as `postbox` in enqueued on this `$screen`.

**Comment by ramiabraham** [comment]
@pippinsplugins 

- Opted to move the `postbox` enqueue to the scripts.php file (https://github.com/AffiliateWP/AffiliateWP/pull/1420/commits/3dd69815736bcaf61489cd8b9933a3f2feb4ac81).

- Traced the `require` error to the `WP_CLI` wrapper in use in the main plugin file:

```php
if ( is_admin() || ( defined( 'WP_CLI' ) && WP_CLI ) ) {
}
```

Placing it in there causes a fatal. Moved it to its' own `is_admin` check just above that (https://github.com/AffiliateWP/AffiliateWP/pull/1420/commits/f83e5bababa669c5e3ff90b3a99de0018bf342b5).

**Comment by pippinsplugins** [comment]
@ramiabraham why does it trigger a fatal when inside the main is_admin() / CLI check? What's the error?

**Comment by ramiabraham** [comment]
@pippinsplugins Class not found; wondering what I'm missing. I'm using a temporary dev environment at the moment (MAMP), so it doesn't have wp-cli, but I also don't see anything wrong with the logic in there that should cause the error, do you?

**Comment by pippinsplugins** [comment]
Can you give me the complete error message?

**Comment by ramiabraham** [comment]
`PHP Fatal error:  Class 'AffWP\Meta_Box\Base' not found in /$PATH/plugins/affiliate-wp/includes/admin/overview/metaboxes/class-metabox-overview-totals.php on line 15`

**Comment by DrewAPicture** [comment]
@ramiabraham I think it's because you need to fully qualify the namespace on the super class in the class declaration.

**Comment by DrewAPicture** [comment]
The fatal is fixed in 9cf1586. We were previously attempting to load overview.php *before* the `Base` class.

**Comment by ramiabraham** [comment]
@pippinsplugins @DrewAPicture noticed the issue, turns out I'm an idiot (https://github.com/AffiliateWP/AffiliateWP/pull/1420/commits/9cf1586541b2b5bdb3bac5cfc7b3626874da1977).

**Comment by pippinsplugins** [comment]
Ah! Good stuff.

**Comment by pippinsplugins** [comment]
@ramiabraham I'm tested this out further by creating a basic sample plugin that adds a new metabox:
```php
<?php

/*
 * Plugin Name: AffiliateWP - Sample Meta Box
 *
*/
namespace AffWP\Meta_Box;

class Test_Affwp_Metabox extends Base {

	public $meta_box_id   = 'my_integration_affwp_metabox';

	public $meta_box_name = 'My Integration AffWP Meta box';
	public $affwp_screen  = 'affiliates_page_affiliate-wp-reports';

	public function content() {
		$this->my_meta_box_content();
	}

	public function my_meta_box_content() {
		echo 'Here is some content I\'d like to share with AffiliateWP users!';
	}
}
new Test_Affwp_Metabox;
```

When I do that, I get this:
```

Fatal error: Class AffWP\Meta_Box\Test_Affwp_Metabox contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (AffWP\Meta_Box\Base::init) in /Users/pippin/Sites/edd/wp-content/plugins/affiliate-wp-sample-metabox/affiliate-wp-sample-metabox.php on line 23
```

I used the sample given at the top of the main meta box class. That sample should be updated to:
```php
namespace AffWP\Meta_Box;

class Test_Affwp_Metabox extends Base {


	public function init() {

		$this->meta_box_id   = 'my_integration_affwp_metabox';

		$this->meta_box_name = 'My Integration AffWP Meta box';
		$this->affwp_screen  = 'affiliates_page_affiliate-wp-reports';

	}

	public function content() {
		$this->my_meta_box_content();
	}

	public function my_meta_box_content() {
		echo 'Here is some content I\'d like to share with AffiliateWP users!';
	}
}
new Test_Affwp_Metabox;
```

That still, however, has an issue because it triggers the following: 
```

Fatal error: Class 'AffWP\Meta_Box\Base' not found in /Users/pippin/Sites/edd/wp-content/plugins/affiliate-wp-sample-metabox/affiliate-wp-sample-metabox.php on line 9
```

Looks like we need to load the class earlier, or provide sample code that uses `plugins_loaded` or later. Could you work up a proper sample that works fully?

**Comment by ramiabraham** [comment]
Added some adjustments (https://github.com/AffiliateWP/AffiliateWP/commit/553e8665226b4a7ea9f0cf765ab4f52f03b7a7ba); the primary issue was the `action` not being defined. Updated the inline doc and example to reflect this as well.

----

As a stand-alone plugin: https://gist.github.com/ramiabraham/1a9d8a09e848a677332c28f7271d5f75

**Comment by ramiabraham** [comment]
@pippinsplugins Looking ok here?

**Comment by pippinsplugins** [comment]
Woot! Working now:
![screen shot 2016-07-25 at 11 40 54 am](https://cloud.githubusercontent.com/assets/1034109/17109209/abea25d0-525c-11e6-8d72-8401810f6d96.png)

**Comment by pippinsplugins** [comment]
@ramiabraham unless I'm wrong, the only screen that currently supports custom metaboxes is the `Overview` screen.

We need to either add the necessary `do_meta_boxes` calls to the other screens or remove the other screen IDs from the base class. We don't want people thinking they can add meta boxes to the other screens when in reality they cannot.

**Comment by ramiabraham** [comment]
@AffiliateWP/core Where do we want meta boxes aside from the `Overview` screen? They're used in the `Reports` screen tabs, but that's it right now.

**Comment by DrewAPicture** [comment]
@ramiabraham What's involved with making other pages compatible? Seems like wrapping the nonces output into a callback that's hooked to a core action would be a reasonably forward-compat method of adding support to other pages – simply fire the action wherever needed and output the `do_meta_boxes()` calls, unless I'm missing something.

**Comment by ramiabraham** [comment]
@DrewAPicture Works for me, good idea.

**Comment by pippinsplugins** [comment]
I'm inclined to say we should do all screens if we do anything, so yeah, let's do them all.

**Comment by ramiabraham** [comment]
Added a function into `misc-functions.php` in https://github.com/AffiliateWP/AffiliateWP/pull/1420/commits/1b90940595239b96cbb929e91770d795a9f9567f which adds global postbox nonces to all affwp admin pages.

**Comment by pippinsplugins** [comment]
@ramiabraham Fairly sure that could result in `Headers already sent` errors since headers haven't been set during `admin_init`. We'll need to use a later hook that adds the nonce inside of our main form elements.

**Comment by ramiabraham** [comment]
(corrected in https://github.com/AffiliateWP/AffiliateWP/commit/173433b08963f3bc1ab4fe140d975fd58048e11a)

**Comment by ramiabraham** [comment]
- Merged into `release/1.9`.
- `issue/1225` left open for unit test coverage additions.

**Comment by DrewAPicture** [comment]
Some minor architectural changes were made to the base class and how core implements meta boxes in #1525, namely:

* Namespaces were adjusted to `AffWP\Admin\Overview\Meta_Box\{Name}` for components (meta boxes), with a top-level namespace of `AffWP\Admin\Meta_Box`
* In order to make instantiating standalone meta boxes possible, the `abstract` keyword was taken off of the top-level `Meta_Box` class and instead all core meta boxes were made to implement a new `AffWP\Admin\Meta_Box\Base` interface, which still allows us to enforce the promise of sub-classing an abstract class, but still allows extending accessing the super-class directly.

```php
namespace AffWP\Admin\Overview\Meta_Box;

use AffWP\Admin\Meta_Box;

/**
 * Implements a Most Valuable Affiliates meta box for the Overview screen.
 *
 * The meta box displays the "most valuable" affiliates, determined by showing the highest:
 *
 * - Earnings
 * - Referrals generated
 * - Visits generated
 *
 * @since 1.9
 * @see   \AffWP\Admin\Meta_Box
 */
class Most_Valuable_Affiliates extends Meta_Box implements Meta_Box\Base {
```
